File: experiments/gilded-rose/python/attempts/opus46high/5/gilded_rose.py
class ItemUpdater:
    def update(self, item):
        self._update_quality(item)
        item.sell_in -= 1
        if item.sell_in < 0:
            self._update_quality_expired(item)

    def _update_quality(self, item):
        item.quality = max(0, item.quality - 1)

    def _update_quality_expired(self, item):
        item.quality = max(0, item.quality - 1)


class AgedBrieUpdater(ItemUpdater):
    def _update_quality(self, item):
        item.quality = min(50, item.quality + 1)

    def _update_quality_expired(self, item):
        item.quality = min(50, item.quality + 1)


class BackstagePassUpdater(ItemUpdater):
    def _update_quality(self, item):
        if item.sell_in < 6:
            item.quality = min(50, item.quality + 3)
        elif item.sell_in < 11:
            item.quality = min(50, item.quality + 2)
        else:
            item.quality = min(50, item.quality + 1)

    def _update_quality_expired(self, item):
        item.quality = 0


class SulfurasUpdater(ItemUpdater):
    def update(self, item):
        pass


class ConjuredUpdater(ItemUpdater):
    def _update_quality(self, item):
        item.quality = max(0, item.quality - 2)

    def _update_quality_expired(self, item):
        item.quality = max(0, item.quality - 2)


def _updater_for(item):
    if item.name == "Aged Brie":
        return AgedBrieUpdater()
    if item.name == "Sulfuras, Hand of Ragnaros":
        return SulfurasUpdater()
    if item.name == "Backstage passes to a TAFKAL80ETC concert":
        return BackstagePassUpdater()
    if item.name.startswith("Conjured"):
        return ConjuredUpdater()
    return ItemUpdater()


class GildedRose:
    def __init__(self, items):
        self.items = items

    def update_quality(self):
        for item in self.items:
            _updater_for(item).update(item)


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)

File: experiments/gilded-rose/csharp/attempts/opus46high/5/GildedRose.cs
using System;
using System.Collections.Generic;

namespace GildedRose;

public class GildedRoseShop
{
    private readonly IList<Item> _items;

    public GildedRoseShop(IList<Item> items)
    {
        _items = items;
    }

    public void UpdateQuality()
    {
        foreach (var item in _items)
            GetUpdater(item).Update(item);
    }

    private static IItemUpdater GetUpdater(Item item) => item.Name switch
    {
        "Aged Brie" => new AgedBrieUpdater(),
        "Backstage passes to a TAFKAL80ETC concert" => new BackstagePassUpdater(),
        "Sulfuras, Hand of Ragnaros" => new SulfurasUpdater(),
        _ when item.Name.StartsWith("Conjured", StringComparison.Ordinal) => new ConjuredUpdater(),
        _ => new NormalUpdater()
    };
}

public interface IItemUpdater
{
    void Update(Item item);
}

public class NormalUpdater : IItemUpdater
{
    public virtual void Update(Item item)
    {
        Degrade(item);
        item.SellIn--;
        if (item.SellIn < 0)
            Degrade(item);
    }

    protected virtual void Degrade(Item item)
    {
        if (item.Quality > 0)
            item.Quality--;
    }
}

public class ConjuredUpdater : NormalUpdater
{
    protected override void Degrade(Item item)
    {
        if (item.Quality > 0) item.Quality--;
        if (item.Quality > 0) item.Quality--;
    }
}

public class AgedBrieUpdater : IItemUpdater
{
    public void Update(Item item)
    {
        IncreaseQuality(item);
        item.SellIn--;
        if (item.SellIn < 0)
            IncreaseQuality(item);
    }

    private static void IncreaseQuality(Item item)
    {
        if (item.Quality < 50)
            item.Quality++;
    }
}

public class BackstagePassUpdater : IItemUpdater
{
    public void Update(Item item)
    {
        if (item.Quality < 50)
        {
            item.Quality++;
            if (item.SellIn < 11 && item.Quality < 50) item.Quality++;
            if (item.SellIn < 6 && item.Quality < 50) item.Quality++;
        }

        item.SellIn--;

        if (item.SellIn < 0)
            item.Quality = 0;
    }
}

public class SulfurasUpdater : IItemUpdater
{
    public void Update(Item item) { }
}
